### PR TITLE
[Need Discussion] Implement twice backward of ConvNd

### PR DESCRIPTION
**Pull request by caogang**
Hi, all.

This PR is following the PR #1555 with cleaner rebase.

@apaszke I have followed your suggestion to modify the `ConvForward` directly. And change the implementation methods of `ConvTransposeNd` to use `ConvBackward` directly instead of `ConvForward(transposed)`. Here is the commit. And I have used `gradcheck` to check the two gradient. It passed the following test.

```python
from torch.autograd import gradcheck
input = (autograd.Variable(torch.randn(1, 20, 22), requires_grad=True),)
test = gradcheck(nn.Conv1d(20, 12, 4), input, eps=1e-3, atol=1e-4)
print(test)
input = (autograd.Variable(torch.randn(1, 20, 22), requires_grad=True),)
test = gradcheck(nn.ConvTranspose1d(20, 12, 4), input, eps=1e-3, atol=1e-4)
print(test)
```

However when I use it to perform grad of grad of `ConvForward`, it raise error. So for now, do not merge this PR.

Assume the forward formula to calculate, is ConvForward(forward_mode) -> ConvForward(forward_mode) -> ConvBackward(grad_mode) ->ConvBackward(grad_mode). And the backward process is ConvForward(grad_mode) -> ConvForward(grad_mode) -> ConvBackward(grad_mode) -> ConvBackward (grad_mode). It raised a `Segment Fault` Error at the **third backward process unit,ConvBackward(grad_mode)**

I have found the problem is because of the `std::move(convolution)` in `ConvBackward::apply`. During the forward process of `ConvBackward(grad_mode)`, it will use `std::move(convolution)` to pass arguments, and `convolution` of itself will be zero. When the second time processing this unit in backward process, some parameters like `convolution` will cause this `Segment error`.

- [x] So how can I pass this `convolution` parameters with itself retained not be cleared when using `std::move` in the following place? 
```python
file : torch/csrc/autograd/functions/convolution.cpp
568+            std::move(columns), std::move(ones), std::move(convolution));
```
- [ ] And the problem remained in PR #1555 

> I have a problem. Assume the forward formula to calculate, is ConvForward -> ConvForward -> ConvBackward ->ConvBackward. And the backward process is ConvForward -> ConvForward -> ConvBackward -> ConvBackward. Should I accumulate grad of weight and bias in every backward process unit? Or just do that in the ConvBackward unit.

### CODE REVIEW

**Comment by apaszke** [comment]
I don't understand what these ConvForward and ConvBackward sequences mean. I also don't really know where do you want to accumulate the grads, these functions only compute and return them, there's no accumulation going on.

**Comment by apaszke** [comment]
Also, you're not doing `gradcheck(ConvNd(20, 12, 4), ...)` right? It's not valid you should never reuse a Function object. This is correct: `gradcheck(lambda input: ConvNd(20, 12, 4)(input), ...)`.

**Comment by caogang** [comment]
Ok, I change the test code as follows. it can still pass the test

```python
inputs = (autograd.Variable(torch.randn(1, 12, 12).double(), requires_grad=True),
               autograd.Variable(torch.randn(10, 12, 5).double(), requires_grad=True,),
               autograd.Variable(torch.randn(10).double(), requires_grad=True))
test = gradcheck(lambda i, w, b : F.conv1d(i, w, b), inputs, eps=1e-6, atol=1e-4)
print(test)
inputs = (autograd.Variable(torch.randn(1, 12, 12).double(), requires_grad=True),
               autograd.Variable(torch.randn(12, 10, 5).double(), requires_grad=True,),
               autograd.Variable(torch.randn(10).double(), requires_grad=True))
test = gradcheck(lambda i, w, b : F.conv1d(i, w, b), inputs, eps=1e-6, atol=1e-4)
print(test)
```

**Comment by caogang** [comment]
By the way. Do you have any idea of the first problem that I want to use another function when passing arguments `convolution` instead of `std::move(convolution)`? Because the `move` will reset the `convolution` parameter, and I want to retain `convolution` value. I have tried pass the `convolution` without `std::move`, it raise error. :(

**Comment by apaszke** [comment]
You need to remove `std::move`, but can't because it's a `unique_ptr` (i.e. there can be only one owner of this pointer). It needs to be changed to a `shared_ptr`.

**Comment by caogang** [comment]
Ok, the first problem is solved. Now I found there are something wrong with accumulating the `grad_weight`  into the `SavedVariable weight_`. So how can I accumulate the grad to `weight_`? Maybe using the `Variable::get_grad_accumulator()`? But I don't know how to use it. @apaszke

**Comment by apaszke** [comment]
You don't need to accumulate any grads! Just return them from `apply` and autograd will take care of the rest

File: torch/csrc/autograd/functions/convolution.cpp [hunk 1]
```diff
@@ -120,204 +143,432 @@ static auto view3d(const Tensor& tensor) -> std::unique_ptr<Tensor> {
...
-    } else {
-      tensor_list outputs(groups);
-      for (int g = 0; g < groups; ++g) {
-        auto input_g = subtensor(input.get(), 1, groups, g);
-        auto weight_g = subtensor(weight.get(), 0, groups, g);
-        auto bias_g = subtensor(bias.get(), 0, groups, g);
-        outputs[g] = compute_output(
-            input_g.get(), weight_g.get(), bias_g.get(),
-            columns[g].get(), ones[g].get(), kernel_size, *this);
+      std::cout << "ConvForward in grad mode" << std::endl;
```
Comment (apaszke) [inline torch/csrc/autograd/functions/convolution.cpp]: Wait, why are there two modes in the forward function?

File: torch/nn/functional.py [hunk 2]
```diff
@@ -155,7 +156,7 @@ def conv_transpose3d(input, weight, bias=None, stride=1, padding=0,
           the number of groups
         dilation: the spacing between kernel elements. Default: 1
     """
-    f = ConvNd(_triple(stride), _triple(padding), _triple(dilation), True,
+    f = ConvNdBackward(_triple(stride), _triple(padding), _triple(dilation), True,
```
Comment (apaszke) [inline torch/nn/functional.py]: It was ok and shouldn't be changed

**Review by apaszke — COMMENTED** [review]
Ok, so @albanD pointed out that grad of grad of ConvNd is not a plain forward convolution, but an expression that has 2x forward convolution and a few additions (I haven't verified that yet though). If that's the case,  then you shouldn't create a `backward` and `forward` mode in the `ConvForward` function, but create `ConvBackwardBackward` which doesn't unpack the Variables in its `apply`, but uses `ConvForward` and `Add` functions to compute it's output. This will be enough to make it infinitely many times differentiable

File: torch/csrc/autograd/functions/convolution.cpp (hunk 1 above)
Comment (caogang) [inline torch/csrc/autograd/functions/convolution.cpp]: Because the convForward in forward mode which means created by the python module and convForward in grad mode which means created by ConvBackward will perform different actions

**Comment by caogang** [comment]
albanD has implemented this feature. Please refer to #1643 .
